include/onnxruntime/core/session/onnxruntime_ep_c_api.h

```diff
@@ -308,6 +309,94 @@ struct OrtKernelImpl {
+   *
+   * \snippet{doc} snippets.dox OrtStatus Return Value
+   *
+   * \note Implementation of this function is optional. If not implemented (set to NULL), ORT assumes the kernel
+   *       does not pre-pack weight data (i.e., `is_packed` defaults to false).
+   *
+   * \since Version 1.24.
+   */
+  ORT_API2_STATUS(PrePackWeight, _In_ OrtKernelImpl* this_ptr, _In_ const OrtValue* tensor,
+                  _In_ int input_index, _Inout_ OrtAllocator* allocator,
+                  _In_opt_ OrtSharedPrePackedWeightCache* prepacked_weight_cache, _Out_ bool* is_packed);
+
+  /** \brief Optional function that receives data for a shared pre-packed weight from ORT.
+   *
+   * This function is called after a prior call to OrtKernelImpl::PrePackWeight set `is_packed` to true and
+   * stored weight data (that the kernel wants to share) into the provided `OrtSharedPrePackedWeightCache` instance.
+   * If this function is called for an unexpected shared weight (as identified by `input_index`), the
+   * implementation should return an error OrtStatus with an error code of ORT_EP_FAIL.
+   * Refer to the description of the "sharing-mode" in the documentation for OrtKernelImpl::PrePackWeight().
+   *
+   * \note This function is based on the internal OpKernel::UseSharedPrePackedBuffers() virtual function used
+   *       within ORT.
+   *
+   * \param[in] this_ptr The OrtKernelImpl instance.
+   * \param[in] buffer_data_ptrs An array of buffer data pointers that collectively hold the pre-packed data for a
+   *                             single shared weight. Note that sometimes a single weight may have multiple pre-packed
+   *                             buffers and it is up to the kernel implementation to determine how to split the data
+   *                             into multiple buffers (if desired) in the call to OrtKernelImpl::PrePack(). Each
+   *                             buffer element in this array is provided in the same order defined by the kernel
+   *                             implementation in the call to OrtKernelImpl::PrePack().
+   * \param[in] buffer_data_sizes An array of buffer byte sizes, one per element in `buffer_data_ptrs`.
+   * \param[in] num_buffers The number of buffers used to store the data for the shared pre-packed weight.
+   *                        Specifies the number of elements in the `buffer_data_ptrs` and `buffer_sizes` arrays.
+   * \param[in] input_index The input index of the tensor in this kernel. This index identifies the identity of
+   *                        the weight.
+   *
+   * \snippet{doc} snippets.dox OrtStatus Return Value
+   *
+   * \note Implementation of this function is generally optional. It is only required if OrtKernelImpl::PrePack()
+   *       elects to share pre-packed weights.
+   *
+   * \since Version 1.24.
+   */
+  ORT_API2_STATUS(SetSharedPrePackedWeight, _In_ OrtKernelImpl* this_ptr,
+                  _In_reads_(num_buffers) const void* const* buffer_data_ptrs,
```

Comment:
Another difference between this function and the internal [OpKernel::UseSharedPrePackedBuffers](https://github.com/microsoft/onnxruntime/blob/19ee5716092ce5cbf8ac5cb8a89bf587a1c123c0/include/onnxruntime/core/framework/op_kernel.h#L126-L128) is that the provided `buffer_data_ptrs` are fully const. This makes sense because the data is now owned by ORT, it may be shared with other kernels, and thus should not be modified by the kernel.